src/components/Rating/Rating.tsx

```diff
@@ -0,0 +1,330 @@
+    // color of an active star
+    color2?: string;
+    half?: boolean;
+    edit?: boolean;
+  };
+  labelIndex: number;
+}
+
+const parentStyles: CSS.Properties = {
+  overflow: 'hidden',
+  // position: 'relative'
+}
+
+const defaultStyles: CSS.Properties = {
+  position: 'relative',
+  // overflow: 'hidden',
+  cursor: 'pointer',
+  display: 'block',
+  float: 'left'
+}
+
+const getHalfStarStyles = (color: any, uniqueness: any) => {
+  return `
+    .react-stars-${uniqueness}:before {
+      position: absolute;
+      overflow: hidden;
+      display: block;
+      z-index: 1;
+      top: 0; left: 0;
+      width: 50%;
+      content: attr(data-forhalf);
+      color: ${color};
+  }`
+}
+
+class Rating extends Component<propTypes, stateTypes> {
+
+  static defaultProps = {
+    edit: true,
+    half: false,
+    value: 0,
+    count: 5,
+    char: '★',
+    size: 30,
+    color1: 'gray',
+    color2: 'orange',
+  
+    onChange: () => { },
+    labels: [],
+    labelIndex: 0
+  };
+
+  constructor(props:propTypes) {
+
+    super(props)
+
+    // set defaults
+
+    props = Object.assign({}, props)
+
+    this.state = {
+      uniqueness: (Math.random() + '').replace('.', ''),
+      value: props.value || 0,
+      stars: [],
+      halfStar: {
+        at: Math.floor(props.value),
+        hidden: props.half && props.value % 1 < 0.5
+      },
+      config: {
+        count: props.count,
+        size: props.size,
+        char: props.char,
+        // default color of inactive star
+        color1: props.color1,
+        // color of an active star
+        color2: props.color2,
+        half: props.half,
+        edit: props.edit,
+      },
+      labelIndex: props.labelIndex
+    }
+
+  }
+
+  componentDidMount() {
+    this.setState({
+      stars: this.getStars(this.state.value)
+    })
+  }
+
+  componentWillReceiveProps(props: propTypes) {
+    this.setState({
+      stars: this.getStars(props.value),
+      value: props.value,
+      halfStar: {
+        at: Math.floor(props.value),
+        hidden: this.state.config.half && props.value % 1 < 0.5
+      },
+      config: Object.assign({}, this.state.config, {
+        count: props.count,
+        size: props.size,
+        char: props.char,
+        color1: props.color1,
+        color2: props.color2,
+        half: props.half,
+        edit: props.edit,
+      })
+    })
+  }
+
+  isDecimal(value: number) {
+    return value % 1 !== 0
+  }
+
+  getRate() {
+    let stars
+    if (this.state.config.half) {
+      stars = Math.floor(this.state.value)
+    } else {
+      stars = Math.round(this.state.value)
+    }
+    return stars
+  }
+
+  getStars(activeCount: number | undefined) {
+    if (typeof activeCount === 'undefined') {
+      activeCount = this.getRate()
+    }
+    let stars = []
+    for (let i = 0; i < this.state.config.count; i++) {
+      stars.push({
+        active: i <= activeCount - 1
+      })
+    }
+    return stars
+  }
+
+  mouseOver(event: any) {
```

Comment:
There is a bug where we hover the target star, the left side of filled stars also increase the size. The preview would look much better if the hovering animation effect is only triggered on the target star. I am happy to help if you need.

Reply:
done

Reply:
I can see the animation effect is only triggered at the target star now. However, there is another bug where the target star is flicking when you are hovering on it. Try hovering through the stars quickly and you will notice the issue. It might be caused by css styling where you may not add transition to the size change?

Reply:
Okay, I'll added transition and replace all exclamation to as keyword.

Reply:
The flicking is still obvious when hovering at large size stars. You may need to check the hovering logic to see if there is anything you can optimize.

Reply:
I have fixed this problem by using scale().

Reply:
Great! Using scale instead is a better solution.